### PR TITLE
[SPARK-28213][SQL] Replace ColumnarBatchScan with equivilant from Columnar

### PR DESCRIPTION
**Pull request by revans2**
## What changes were proposed in this pull request?

This is a second part of the https://issues.apache.org/jira/browse/SPARK-27396 and a follow on to #24795 

## How was this patch tested?

I did some manual tests and ran/updated the automated tests

I did some simple performance tests on a single node to try to verify that there is no performance impact, and I was not able to measure anything beyond noise.

### CODE REVIEW

**Comment by tgravescs** [comment]
add to whitelist

**Comment by tgravescs** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala [hunk 1]
```diff
@@ -108,58 +108,62 @@ case class InMemoryTableScanExec(
...
+      columnarIterator.initialize(withMetrics, columnTypes, requestedColumnIndices.toArray)
+      if (enableAccumulatorsForTest && columnarIterator.hasNext) {
+        readPartitions.add(1)
       }
+      columnarIterator
     }
   }
 
-  override def inputRDDs(): Seq[RDD[InternalRow]] = Seq(inputRDD)
+  // override def inputRDDs(): Seq[RDD[InternalRow]] = Seq(inputRDD)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: Shall we remove this cleanly?

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 2]
```diff
@@ -1292,8 +1292,10 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
       val df = sql("SELECT * FROM a WHERE p <= (SELECT MIN(id) FROM b)")
       checkAnswer(df, Seq(Row(0, 0), Row(2, 0)))
       // need to execute the query before we can examine fs.inputRDDs()
+      df.explain
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: Shall we clean up this?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala [hunk 3]
```diff
@@ -485,16 +485,16 @@ class InMemoryColumnarQuerySuite extends QueryTest with SharedSQLContext {
     // Push predicate to the cached table.
     val df2 = df1.where("y = 3")
 
+    logWarning(s"ORIG QUERY PLAN:\n${df2.queryExecution.executedPlan}")
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala]: Ur, shall we clean up these three `logWarning`s in this test suite? This will not be read.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala [hunk 4]
```diff
@@ -108,59 +108,61 @@ case class InMemoryTableScanExec(
...
   }
 
+  private lazy val columnarInputRDD: RDD[ColumnarBatch] = {
+    val numOutputRows = longMetric("numOutputRows")
+    val buffers = filteredCachedBatches()
+    val offHeapColumnVectorEnabled = conf.offHeapColumnVectorEnabled
+    buffers
+      .map(createAndDecompressColumn(_, offHeapColumnVectorEnabled))
+      .map(b => {
+        numOutputRows += b.numRows()
```
Comment (abellina) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: should `numOutputRows` be `max(numRows)`?

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 5]
```diff
@@ -1293,7 +1293,8 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
       checkAnswer(df, Seq(Row(0, 0), Row(2, 0)))
       // need to execute the query before we can examine fs.inputRDDs()
       assert(df.queryExecution.executedPlan match {
-        case WholeStageCodegenExec(fs @ FileSourceScanExec(_, _, _, partitionFilters, _, _, _)) =>
+        case WholeStageCodegenExec(ColumnarToRowExec(InputAdapter(
```
Comment (abellina) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: Ok, so `ColumnarToRowExec` is here because `InputAdapter` may `supportColumnar`, right? Why is `InputAdapter` getting added here?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala [hunk 6]
```diff
@@ -121,29 +121,6 @@ class WholeStageCodegenSuite extends QueryTest with SharedSQLContext {
     assert(ds.collect() === Array(("a", 10.0), ("b", 3.0), ("c", 1.0)))
   }
 
-  test("cache for primitive type should be in WholeStageCodegen with InMemoryTableScanExec") {
```
Comment (abellina) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala]: Not clear to me why this test doesn't apply anymore.

**Review by abellina — COMMENTED** [review]
Overall it makes sense. Just had some high level questions.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala (hunk 4 above)
Comment (revans2) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: no because `b` is a `ColumnarBatch`, so we are iterating over possibly multiple batches.  We are not iterating over individual columns.

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala (hunk 5 above)
Comment (revans2) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: WholeStageCodeGen marks the end of a code generation stage.  InputAdapter marks the beginning of a code generation stage.  So what we had before was a WholeStageCodeGen that had it's first entry a FileSourceScanExec because before this change FileSourceScanExec supported code generation to convert ColumnarBatchs into rows.  The InputAdapter would logically have been a child of FileSourceScanExec, but it has no children so it is not there.

After this change ColumnarToRowExec is the only thing in the code generation stage, so it is flanked by the WholeStageCodegenExec and the InputAdaptor.  FileSourceScanExec is returning batches and is not doing code gen because it is not needed any longer.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala (hunk 6 above)
Comment (revans2) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala]: It was verifying that InMemoryTableScanExec was inside of a WholeStageCodegen phase, but after this change InMemoryTableScanExec no longer supports codegen so the test is invalid.  The ColumnarToRowExec is what will be in the codegen section instead.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala (hunk 4 above)
Comment (abellina) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: ah right, missed that. Thanks

**Review by tgravescs — APPROVED** [review]
changes look good.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 7]
```diff
@@ -141,11 +148,11 @@ case class FileSourceScanExec(
     optionalBucketSet: Option[BitSet],
     dataFilters: Seq[Expression],
     override val tableIdentifier: Option[TableIdentifier])
-  extends DataSourceScanExec with ColumnarBatchScan  {
+  extends DataSourceScanExec {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Is this change making all `DataSourceScanExec` nodes not codegen support, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala [hunk 8]
```diff
@@ -96,9 +96,6 @@ case class ColumnarToRowExec(child: SparkPlan)
   /**
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala]: `ColumnarToRowExec`'s comment also mentions `ColumnarBatchScan`. If you are like to remove all reference to `ColumnarBatchScan`...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala [hunk 9]
```diff
@@ -24,17 +24,20 @@ import org.apache.spark.sql.catalyst.dsl.expressions._
...
 import org.apache.spark.sql.catalyst.plans.QueryPlan
 import org.apache.spark.sql.catalyst.plans.physical.Partitioning
-import org.apache.spark.sql.execution.{ColumnarBatchScan, LeafExecNode, SparkPlan, WholeStageCodegenExec}
+import org.apache.spark.sql.execution.{LeafExecNode, SparkPlan, WholeStageCodegenExec}
+import org.apache.spark.sql.execution.metric.SQLMetrics
 import org.apache.spark.sql.execution.vectorized._
 import org.apache.spark.sql.types._
 import org.apache.spark.sql.vectorized.{ColumnarBatch, ColumnVector}
 
-
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: It would be good to avoid unnecessary change like this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 7 above)
Comment (revans2) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Correct, but there were only 2 things that the code gen was doing. Either convert ColumnarBatch into UnsafeRows or to convert whatever other rows were being returned by the DataSourceScanExec into UnsafeRows.  The ColumnarBatch conversion is now covered by ColumnarToRowExec.  The row to row conversion is covered by UnsafeProjections that are either inserted as a part of this patch or were already in the code, so we ended up doing a double conversion.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala (hunk 8 above)
Comment (revans2) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala]: Great catch. I thought I got rid of all of them.  Will grep though again.

**Review by tgravescs — APPROVED** [review]
+1

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala [hunk 10]
```diff
@@ -392,7 +392,7 @@ private[sql] trait SQLTestUtilsBase
    */
   protected def stripSparkFilter(df: DataFrame): DataFrame = {
     val schema = df.schema
-    val withoutFilters = df.queryExecution.sparkPlan.transform {
+    val withoutFilters = df.queryExecution.executedPlan.transform {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala]: Why making this change in this PR?

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala]: Seems like the existing columnar logic within each plans have moved to `RowToColumnarExec` and `ColumnarToRowExec` to deduplicate but it's now dependent on `ApplyColumnarRulesAndInsertTransitions` rule, which, now, requires execution preparation (`QueryExecution.preparations`).

However, per the doc, `executedPlan` should be only used for execution ideally. It could have been best to avoid. @revans2 even though it's too late, can you please describe what does this PR fixes in the PR description (presumably by listing each item)?

I have no idea what this PR fixes from reading the PR description and JIRA.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/LogicalPlanTagInSparkPlanSuite.scala [hunk 11]
```diff
@@ -44,9 +44,14 @@ class LogicalPlanTagInSparkPlanSuite extends TPCDSQuerySuite {
...
 
   // A scan plan tree is a plan tree that has a leaf node under zero or more Project/Filter nodes.
-  private def isScanPlanTree(plan: SparkPlan): Boolean = plan match {
-    case p: ProjectExec => isScanPlanTree(p.child)
-    case f: FilterExec => isScanPlanTree(f.child)
+  // Because of how codegen and columnar to row transitions work, we may have InputAdaptors
+  // and ColumnarToRow transformations in the middle of it, but they will not have the tag
+  // we want, so skip them if they are the first thing we see
+  private def isScanPlanTree(plan: SparkPlan, first: Boolean): Boolean = plan match {
+    case i: InputAdapter if !first => isScanPlanTree(i.child, false)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/LogicalPlanTagInSparkPlanSuite.scala]: Why is `InputAdapter` pop up?

**Comment by HyukjinKwon** [comment]
Hey, I see roughly the core logic itself is deduplicated fine without changing the exiting codes itself but seems like it changes the other stuff.

Let's be clear on what this PR fixes next time - I thought this was just a simple refactoring but now realised that this is actually pretty invasive.

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala (hunk 10 above)
Comment (revans2) [reply sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala]: This PR replaces ColumnarBatchScan with ColumnarToRowExec.  This involved changes to all subclasses of COlumnarBatchScan and to AdaptiveSparkPlanExec so it would also execute the needed columnar transition rules.  I also made any fixes needed for tests.  I preferred to keep the changes as small as possible for the tests, which is why I made a small change here in a test utility class.

The issue was that some tests were directly execution the plan returned by this function, which used to work for some very limited use cases, but did not work in all cases.  I am happy to try and fix issues with this approach for the tests, just let me know what is the correct way to do it?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/LogicalPlanTagInSparkPlanSuite.scala (hunk 11 above)
Comment (revans2) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/LogicalPlanTagInSparkPlanSuite.scala]: Prior to this PR, a subclass of ColumanarBatchScan would be in a plan that looked like.

```
INPUT (subclass of ColumnarBatchScan) -> (code gen supported nodes) -> WholeStageCodegenExec -> ...
```

After this change, it now looks like

```
INPUT (not a subclass of ColumnarBatchScan) -> InputAdapter -> ColumnarToRowExec -> (code gen supported nodes) -> WholeStageCodegenExec -> ...
```

Because the INPUT class no longer supports code generation the code generation rule will insert an InputAdapter after it and before the ColumnarToRowExec that does support code generation.

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala (hunk 10 above)
Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala]: I don't know the correct way to fix it for now - it needs some investigations. Can you clarify why it didn't work well?

The problem seems like by the internal behaviour changes - we now rely on `ApplyColumnarRulesAndInsertTransitions`. Can we make some investigation to confirm that it doesn't affect anything, and list up what changes were made in this PR description?

Comment (revans2) [reply sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala]: Yes that is exactly what it is.

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala]: @revans2, what I meant is like https://github.com/apache/spark/pull/25264 rather than  single line that describes what this PR proposes.
